dev/com.ibm.ws.sipcontainer/src/com/ibm/ws/sip/stack/transport/sip/netty/SipUdpConnLink.java

```diff
@@ -0,0 +1,492 @@
+									// measure time duration in sip stack queue
+									PerformanceMgr.getInstance().measureTaskDurationOutboundQueue(
+											messageContext.getSipContainerQueueDuration().takeTimeMeasurement());
+
+									if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+										Tr.debug(this, tc, "run",
+												"update QueueMonitoring outbound queue statistics - task dequeued");
+									}
+									PerformanceMgr.getInstance().updateQueueMonitoringTaskDequeuedFromOutboundQueue();
+								}
+							}
+
+						} catch (NoSuchElementException e) {
+							// according to the Object wait() API: "spurious wakeups are possible"
+							// we saw this issue with SUN JVM, so if we get here and list is empty we just
+							// need
+							// to wait again
+							if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+								Tr.debug(this, tc, "run", "Trying to remove message from empty list");
+							}
+							continue;
+						}
+					}
+					// there's something to send.
+					// wait here until it's safe to send
+					synchronized (this) {
+						if (m_locked) {
+							wait();
+						}
+					}
+					if (!m_running) {
+						break;
+					}
+					// there's something to send, and it's safe to send it
+					m_locked = true;
+					boolean immediate;
+					try {
+						immediate = m_connLink.sendNow(messageContext);
+						messageContext.writeComplete(); // the message context is no longer referenced
+					} catch (IOException e) {
+						if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+							Tr.debug(this, tc, "run", "IOException", e);
+						}
+						messageContext.writeError(e);
+						immediate = true;
+					}
+					if (immediate) {
+						// completed immediately
+						sendComplete();
+					} else {
+						// wait for completion
+					}
+				}
+			} catch (InterruptedException e) {
+				if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+					Tr.debug(this, tc, "run", "InterruptedException", e);
+				}
+			}
+			if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+				Tr.debug(this, tc, "SipUdpConnLink thread terminated");
+			}
+		}
+	}
+
+	/**
+	 * @param channel inbound channel that is creating or accessing this conn link
+	 * @return conn link instance given the listening point
+	 */
+	static SipUdpConnLink instance(SipUdpInboundChannel channel) {
+		ListeningPoint lp = channel.getListeningPoint();
+		SipUdpConnLink connLink = (SipUdpConnLink) s_instances.get(lp);
+		if (connLink == null) {
+			connLink = new SipUdpConnLink(channel);
+			s_instances.put(lp, connLink);
+		}
+		return connLink;
+	}
+
+	/**
+	 * private constructor
+	 * 
+	 * @param channel channel that is creating this connection
+	 */
+	private SipUdpConnLink(SipUdpInboundChannel channel) {
+		m_channel = channel;
+		/*
+		 * TODO Liberty m_needToLearnRouterEndpoint =
+		 * AdminHelper.getPlatformHelper().isZOS();
```

Comment:
TODO?

Reply:
Nope.  z/OS implementation not contemplated at this time.